fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/pdagg/PushDownAggregateCollector.java

```diff
@@ -0,0 +1,550 @@
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CaseWhenOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.statistics.ColumnStatistic;
+import com.starrocks.sql.optimizer.statistics.ExpressionStatisticCalculator;
+import com.starrocks.sql.optimizer.statistics.Statistics;
+import com.starrocks.sql.optimizer.statistics.StatisticsCalculator;
+import com.starrocks.sql.optimizer.statistics.StatisticsEstimateCoefficient;
+import com.starrocks.sql.optimizer.task.TaskContext;
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/*
+ * Collect all can be push down aggregate context, to get which aggregation can be
+ * pushed down and the push down path.
+ *
+ * Can't rewrite directly, because we don't know which aggregation needs to be
+ * push down before arrive at scan node.
+ *
+ * And in this phase, the key of AggregateContext's map is origin aggregate column, for
+ * mark push down which aggregation, the value will multi-rewrite by path, for check
+ * which aggregation needs push down
+ */
+class PushDownAggregateCollector extends OptExpressionVisitor<Void, AggregateContext> {
+    private static final List<String> WHITE_FNS = ImmutableList.of(FunctionSet.MAX, FunctionSet.MIN,
+            FunctionSet.SUM, FunctionSet.AVG, FunctionSet.HLL_UNION, FunctionSet.BITMAP_UNION,
+            FunctionSet.PERCENTILE_UNION);
+
+    private final TaskContext taskContext;
+    private final OptimizerContext optimizerContext;
+    private final ColumnRefFactory factory;
+    private final SessionVariable sessionVariable;
+
+    private final Map<LogicalAggregationOperator, List<AggregateContext>> allRewriteContext = Maps.newHashMap();
+    // private final Map<Integer, List<AggregateContext>> cteRewriteContexts = Maps.newHashMap();
+
+    public PushDownAggregateCollector(TaskContext taskContext) {
+        this.taskContext = taskContext;
+        optimizerContext = taskContext.getOptimizerContext();
+        factory = taskContext.getOptimizerContext().getColumnRefFactory();
+        sessionVariable = taskContext.getOptimizerContext().getSessionVariable();
+    }
+
+    Map<LogicalAggregationOperator, List<AggregateContext>> getAllRewriteContext() {
+        return allRewriteContext;
+    }
+
+    public void collect(OptExpression root) {
+        process(root, new AggregateContext());
+    }
+
+    @Override
+    public Void visit(OptExpression optExpression, AggregateContext context) {
+        // forbidden push down
```

Comment:
Why still need to process child with EMPTY context in this forbidden situation.

Reply:
maybe children is an aggregate node and can be push